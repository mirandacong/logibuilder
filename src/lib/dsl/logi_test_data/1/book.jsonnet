local defs = import '../defs.jsonnet';

local get_column_block(table_name, block_name) = (
    defs.Col('FY', {
        labels: ['strategy', 'CHECK'],
        slice_exprs: [
            defs.Slice(
                'stock',
                '{%s!%s!Q1}+' % [table_name, block_name] +
                '{%s!%s!Q2}+' % [table_name, block_name] +
                '{%s!%s!Q3}+' % [table_name, block_name] +
                '{%s!%s!Q4}' % [table_name, block_name]
            ),
            defs.Slice('current', '{%s!%s!Q4}' % [table_name, block_name],),
        ],
    })
);

local rb_sub_row = [
    defs.Row('iPhone', {
        modifier: '-',
        labels: ['assumption'],
        slice_exprs: [
            defs.Slice('proj', ''),
            defs.Slice('hist', ''),
        ],
    }),
    defs.Row('iPad', {
        modifier: '-',
        slice_exprs: [
            defs.Slice('proj', '{iPad}[proj].lag(1)*({iPad}[hist]/{iPad}[hist].lag(1)).average()'),
            defs.Slice('hist', ''),
        ],
    }),
    defs.Row('Mac', {
        slice_exprs: [
            defs.Slice('proj', ''),
            defs.Slice('hist', ''),
        ],
    }),
    defs.Row('Total', {
        slice_exprs: [
            defs.Slice('hist', '{iPhone}[hist]+{iPad}[hist]+{Mac}[hist]'),
            defs.Slice('proj', '{iPhone}[proj] + {iPad}[proj] + {Mac}[proj]'),
        ],
    }),
];


local table1 = defs.Table('Revenue Breakup', {
    header_stub: 'us mn',
    subnodes: [
        defs.ColBlock('2017', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Revenue Breakup', '2017'),
            ],
        }),
        defs.ColBlock('2018', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Revenue Breakup', '2018'),
            ],
        }),
        defs.ColBlock('2019', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Revenue Breakup', '2019'),
            ],
        }),
        defs.ColBlock('2020', {
            labels: ['proj'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                defs.Col('FY', {}),
            ],
        }),
        defs.Row('Revenue', {
            expression: 'SUM({Electronic products!Total},{Services})',
            labels: ['stock'],
            modifier: '++',
        }),
        defs.Row('yoy growth', {
            expression: '{Revenue Breakup!Revenue}/{Revenue Breakup!Revenue}.lag(1)-1',
            labels: ['Revenue'],
            modifier: '--',
            alias: '1',
        }),
        defs.Row('Electronic products', {
            modifier: '+',
        }),
        defs.RowBlock('Electronic products', {
            modifier: '+',
            subnodes: rb_sub_row,
        }),
        defs.Row('yoy growth', {
            expression: '{Electronic products!Total}',
            labels: ['Total'],
            modifier: '--',
            alias: '2',
        }),
        defs.Row('Services', {
            modifier: '+',
            slice_exprs: [
                defs.Slice('hist', ''),
                defs.Slice('proj', '{Services}[proj].lag(1)*(1 + {Control!Revenue Breakup!yoy growth@3})'),
            ],
        }),
        defs.Row('yoy growth', {
            expression: '',
            is_def_scalar: true,
            labels: ['service_growth', 'assumption'],
            modifier: '--',
            alias: '3',
        }),
    ],
});

local table2 = defs.Table('Income Statement', {
    header_stub: 'us mn',
    subnodes: [
        defs.ColBlock('2017', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Income Statement', '2017'),
            ],
        }),
        defs.ColBlock('2018', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Income Statement', '2018'),
            ],
        }),
        defs.ColBlock('2019', {
            labels: ['hist'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                get_column_block('Income Statement', '2019'),
            ],
        }),
        defs.ColBlock('2020', {
            labels: ['proj'],
            subnodes: [
                defs.Col('Q1', {}),
                defs.Col('Q2', {}),
                defs.Col('Q3', {}),
                defs.Col('Q4', {}),
                defs.Col('FY', {}),
            ],
        }),
        defs.Row('Cost of Sales', {
            labels: ['current'],
            modifier: '+',
            slice_exprs: [
                defs.Slice('hist', ''),
                defs.Slice('proj', ''),
            ],
        }),
        defs.Row('Operating Expenses', {
            labels: ['current'],
            slice_exprs: [
                defs.Slice('hist', ''),
                defs.Slice('proj', '{Revenue Breakup!Revenue} * {as %revenue}[hist].average()'),
            ],
        }),
        defs.Row('as %revenue', {
            expression: '{Operating Expenses}/{Revenue Breakup!Revenue}',
            modifier: '--',
            slice_exprs: [
                defs.Slice('hist', ''),
            ],
        }),
        defs.Row('Operating Income', {
            expression: '{Revenue Breakup!Revenue} - {Operating Expenses}',
        }),
        defs.Row('Total Interest & Other Income', {
            slice_exprs: [
                defs.Slice('hist', ''),
                defs.Slice('proj', ''),
            ],
        }),
        defs.Row('Pretax Income', {
            expression: '{Operating Income} + {Total Interest & Other Income}',
        }),
        defs.Row('tax rate', {
            expression: '',
            is_def_scalar: true,
        }),
        defs.Row('tax', {
            expression: '{Operating Income} * {tax rate}',
        }),
        defs.Row('Net Income', {
            expression: '{Pretax Income} - {tax}',
            modifier: '+',
        }),
        defs.Row('Gross Profit', {
            expression: '{Revenue Breakup!Revenue} - {Cost of Sales}',
        }),
        defs.Row('Gross Margin', {
            expression: '{Gross Profit} / {Revenue Breakup!Revenue}',
            modifier: '--',
        }),
    ],
});

local sheet = defs.Sheet('Control', {
    subnodes: [
        defs.Title('Financial Statement', {
            modifier: '++',
            subnodes: [defs.Title('China', { modifier: '+' })],
        }),
        table1,
        defs.Title('Income Statement', {}),
        table2,
    ],
});

local book = defs.Book('Test model', {
    subnodes: [sheet],
});

book
