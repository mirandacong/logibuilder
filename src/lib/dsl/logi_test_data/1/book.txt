@0,0,0,0@
```(
book Test model
    sheet Control
        ++ title Financial Statement
            + title China
        header period
            stub us mn
            [[hist]] 2017 | [[hist]] 2018 | [[hist]] 2019
                Q1 | Q2 | Q3 | Q4 | FY
                [[priority=6, strategy=CHECK]]FY:
                    [stock] = {Q1} + {Q2} + {Q3} + {Q4}
                    [current] = {Q4}
            [[proj]] 2020
                Q1 | Q2 | Q3 | Q4 | FY

        table Revenue Breakup
            header = {period: Test Model/Control}

            ++ [[stock]]Revenue = SUM({Total:\
                    Test model/Control/Revenue Breakup/Electronic products}, \
                    {Services})

            -- yoy growth[[Revenue]] = {Revenue}/{Revenue}.lag(1)-1

            + Electronic products
                - [[assumption]]iPhone[proj] = input()
                    [hist] = source
                - iPad[proj] ={iPad[proj]}.lag(1)*({iPad[hist]}/{iPad[hist]}.lag(1)).average()
                    [hist] = source
                - Mac:
                    [[assumption]][proj] = input()
                Total:
                    [[strategy=CHECK]][hist] = {iPhone[hist]}+{iPad[hist]}+{Mac[hist]}
                    [proj] = {iPhone[proj]} + {iPad[proj]} + {Mac[proj]}

            -- yoy growth[[Total]] = {Total: Test model/Control/Revenue Breakup/Electronic products}

            + Services:
                [[fact]][hist] = input()
                [proj] = {Services[proj]}.lag(1)*(1 + {yoy growth[[service_growth]]})

            -- @ [[service_growth, assumption]] yoy growth = input()


        title Income Statement
            table Income Statement
                header = {period: Test model/Control}

                + [[current]] Cost of Sales:
                    [[fact]][hist] = source()
                    [[assumption]][proj] = input()

                [[current]] Operating Expenses:
                    [[estimated]][hist] = input()
                    [proj] = {Revenue: Test model/Control/Revenue Breakup}\
                        * {as %revenue[hist]}.average()

                -- as %revenue = {Operating Expenses}/{Revenue:\
                Test model/Control/Revenue Breakup}

                Operating Income = {Revenue:\
                    Test model/Control/Revenue Breakup} \
                    - {Operating Expenses}

                Total Interest & Other Income:
                    [[fact]][hist] = source()
                    [[assumption]][proj] = input()

                Pretax Income = {Operating Income} + {Total Interest & Other Income}

                @ tax rate = input()
                tax = {Operating Income} * {tax rate}

                + Net Income = {Pretax Income} - {tax}



                Gross Profit = {Revenue:\
                    Test model/Control/Revenue Breakup} - {Cost of Sales}
                -- Gross Margin = {Gross Profit} / {Revenue:\
                    Test model/Control/Revenue Breakup}
)```
