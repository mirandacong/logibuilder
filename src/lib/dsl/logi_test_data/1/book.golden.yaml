---
annotations: {}
labels: []
name: Test model
nodetype: 1
subnodes:
- annotations: {}
  labels: []
  name: Control
  nodetype: 2
  subnodes:
  - annotations: {}
    labels: []
    modifier: ++
    name: Financial Statement
    nodetype: 3
    subnodes:
    - annotations: {}
      labels: []
      modifier: +
      name: China
      nodetype: 3
      subnodes: []
  - annotations: {}
    header_stub: us mn
    labels: []
    name: Revenue Breakup
    nodetype: 4
    subnodes:
    - annotations: {}
      labels:
      - hist
      name: '2017'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Revenue Breakup!2017!Q1}+{Revenue Breakup!2017!Q2}+{Revenue
              Breakup!2017!Q3}+{Revenue Breakup!2017!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Revenue Breakup!2017!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - hist
      name: '2018'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Revenue Breakup!2018!Q1}+{Revenue Breakup!2018!Q2}+{Revenue
              Breakup!2018!Q3}+{Revenue Breakup!2018!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Revenue Breakup!2018!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - hist
      name: '2019'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Revenue Breakup!2019!Q1}+{Revenue Breakup!2019!Q2}+{Revenue
              Breakup!2019!Q3}+{Revenue Breakup!2019!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Revenue Breakup!2019!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - proj
      name: '2020'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
    - alias: ''
      annotations: {}
      expression: SUM({Electronic products!Total},{Services})
      is_def_scalar: false
      labels:
      - stock
      modifier: ++
      name: Revenue
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: '1'
      annotations: {}
      expression: '{Revenue Breakup!Revenue}/{Revenue Breakup!Revenue}.lag(1)-1'
      is_def_scalar: false
      labels:
      - Revenue
      modifier: --
      name: yoy growth
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: false
      labels: []
      modifier: +
      name: Electronic products
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - annotations: {}
      labels: []
      modifier: +
      name: Electronic products
      nodetype: 5
      subnodes:
      - alias: ''
        annotations: {}
        expression: ''
        is_def_scalar: false
        labels:
        - assumption
        modifier: '-'
        name: iPhone
        nodetype: 7
        slice_exprs:
        - annotations: {}
          expression: ''
          name: proj
          type: 0
        - annotations: {}
          expression: ''
          name: hist
          type: 0
        sources: []
        subnodes: []
        type: 0
      - alias: ''
        annotations: {}
        expression: ''
        is_def_scalar: false
        labels: []
        modifier: '-'
        name: iPad
        nodetype: 7
        slice_exprs:
        - annotations: {}
          expression: '{iPad}[proj].lag(1)*({iPad}[hist]/{iPad}[hist].lag(1)).average()'
          name: proj
          type: 0
        - annotations: {}
          expression: ''
          name: hist
          type: 0
        sources: []
        subnodes: []
        type: 0
      - alias: ''
        annotations: {}
        expression: ''
        is_def_scalar: false
        labels: []
        modifier: ''
        name: Mac
        nodetype: 7
        slice_exprs:
        - annotations: {}
          expression: ''
          name: proj
          type: 0
        - annotations: {}
          expression: ''
          name: hist
          type: 0
        sources: []
        subnodes: []
        type: 0
      - alias: ''
        annotations: {}
        expression: ''
        is_def_scalar: false
        labels: []
        modifier: ''
        name: Total
        nodetype: 7
        slice_exprs:
        - annotations: {}
          expression: '{iPhone}[hist]+{iPad}[hist]+{Mac}[hist]'
          name: hist
          type: 0
        - annotations: {}
          expression: '{iPhone}[proj] + {iPad}[proj] + {Mac}[proj]'
          name: proj
          type: 0
        sources: []
        subnodes: []
        type: 0
    - alias: '2'
      annotations: {}
      expression: '{Electronic products!Total}'
      is_def_scalar: false
      labels:
      - Total
      modifier: --
      name: yoy growth
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: false
      labels: []
      modifier: +
      name: Services
      nodetype: 7
      slice_exprs:
      - annotations: {}
        expression: ''
        name: hist
        type: 0
      - annotations: {}
        expression: '{Services}[proj].lag(1)*(1 + {Control!Revenue Breakup!yoy growth@3})'
        name: proj
        type: 0
      sources: []
      subnodes: []
      type: 0
    - alias: '3'
      annotations: {}
      expression: ''
      is_def_scalar: true
      labels:
      - service_growth
      - assumption
      modifier: --
      name: yoy growth
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
  - annotations: {}
    labels: []
    modifier: ''
    name: Income Statement
    nodetype: 3
    subnodes: []
  - annotations: {}
    header_stub: us mn
    labels: []
    name: Income Statement
    nodetype: 4
    subnodes:
    - annotations: {}
      labels:
      - hist
      name: '2017'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Income Statement!2017!Q1}+{Income Statement!2017!Q2}+{Income
              Statement!2017!Q3}+{Income Statement!2017!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Income Statement!2017!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - hist
      name: '2018'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Income Statement!2018!Q1}+{Income Statement!2018!Q2}+{Income
              Statement!2018!Q3}+{Income Statement!2018!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Income Statement!2018!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - hist
      name: '2019'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels:
        - strategy
        - CHECK
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs:
        - annotations: {}
          expression: '{Income Statement!2019!Q1}+{Income Statement!2019!Q2}+{Income
              Statement!2019!Q3}+{Income Statement!2019!Q4}'
          name: stock
          type: 0
        - annotations: {}
          expression: '{Income Statement!2019!Q4}'
          name: current
          type: 0
        sources: []
        subnodes: []
        type: 0
    - annotations: {}
      labels:
      - proj
      name: '2020'
      nodetype: 6
      subnodes:
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q1
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q2
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q3
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: Q4
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
      - annotations: {}
        expression: ''
        labels: []
        modifier: ''
        name: FY
        nodetype: 8
        slice_exprs: []
        sources: []
        subnodes: []
        type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: false
      labels:
      - current
      modifier: +
      name: Cost of Sales
      nodetype: 7
      slice_exprs:
      - annotations: {}
        expression: ''
        name: hist
        type: 0
      - annotations: {}
        expression: ''
        name: proj
        type: 0
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: false
      labels:
      - current
      modifier: ''
      name: Operating Expenses
      nodetype: 7
      slice_exprs:
      - annotations: {}
        expression: ''
        name: hist
        type: 0
      - annotations: {}
        expression: '{Revenue Breakup!Revenue} * {as %revenue}[hist].average()'
        name: proj
        type: 0
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Operating Expenses}/{Revenue Breakup!Revenue}'
      is_def_scalar: false
      labels: []
      modifier: --
      name: as %revenue
      nodetype: 7
      slice_exprs:
      - annotations: {}
        expression: ''
        name: hist
        type: 0
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Revenue Breakup!Revenue} - {Operating Expenses}'
      is_def_scalar: false
      labels: []
      modifier: ''
      name: Operating Income
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: false
      labels: []
      modifier: ''
      name: Total Interest & Other Income
      nodetype: 7
      slice_exprs:
      - annotations: {}
        expression: ''
        name: hist
        type: 0
      - annotations: {}
        expression: ''
        name: proj
        type: 0
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Operating Income} + {Total Interest & Other Income}'
      is_def_scalar: false
      labels: []
      modifier: ''
      name: Pretax Income
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: ''
      is_def_scalar: true
      labels: []
      modifier: ''
      name: tax rate
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Operating Income} * {tax rate}'
      is_def_scalar: false
      labels: []
      modifier: ''
      name: tax
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Pretax Income} - {tax}'
      is_def_scalar: false
      labels: []
      modifier: +
      name: Net Income
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Revenue Breakup!Revenue} - {Cost of Sales}'
      is_def_scalar: false
      labels: []
      modifier: ''
      name: Gross Profit
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
    - alias: ''
      annotations: {}
      expression: '{Gross Profit} / {Revenue Breakup!Revenue}'
      is_def_scalar: false
      labels: []
      modifier: --
      name: Gross Margin
      nodetype: 7
      slice_exprs: []
      sources: []
      subnodes: []
      type: 0
